### ISSUE TITLE
Tags set using keywords don't appear in dryrun logs

### ISSUE THREAD

**Issue by lesnake** [robotframework/robotframework#3985]
Tags set using `set tags`, `pass execution`, `pass execution if` don't appear in dryrun log.

Run the following suite two times, using the command:
`robot -d normal tst_passExecTag.robot; robot -d dryrun --dryrun tst_passExecTag.robot`

```
*** settings ***
Documentation   Run with command robot --dryrun tst_passExecTag.robot
...
...    Run with command
...    ``robot -d normal tst_passExecTag.robot``
...    then with command
...    ``robot -d dryrun --dryrun tst_passExecTag.robot``

*** test cases ***
tst
    # This tag wont appear either
    set tags   tag set using set tags
    Pass execution  The following tag will appear in normal run mode but not in dryrun  tag set using pass execution
```

Then open the log.html files and compare their 'statistics by tag" section.

They differ, where I expect them to be the same.

This reduces the information level generated by dryrun.

**Comment by vinayvennela** [comment]
@lesnake  As per the doc, see this about dry run

https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#dry-run

Robot Framework supports so called dry run mode where the tests are run normally otherwise, but the keywords coming from the test libraries are not executed at all. The dry run mode can be used to validate the test data; if the dry run passes, the data should be syntactically correct.

In dry run only syntactical checks are done, the keywords are not at all executed. In your example, you are adding tags as part of the keyword which will not be added to the report as it is not executed and tags are not saved. So in your dryrun/log.html  expand the keywords and you can clearly see that nothing is executed

**Comment by pekkaklarck** [comment]
Yeah, @vinayvennela is right that keywords aren't in general run the dry-run mode and thus actions they should do, like modifying tags by `Set Tags`, aren't done. There are, however, exceptions to this rule and keywords like `Import Library` are actually run. We could make such an exception also with `Set Tags` and `Remove Tags`. Is someone interested to implement such a functionality? Ought to be easy.

**Comment by vinayvennela** [comment]
I'm in. If you could guide me. @pekkaklarck

**Comment by pekkaklarck** [comment]
Cool! The code controlling which keywords are run in dry-run is here:
https://github.com/robotframework/robotframework/blob/master/src/robot/running/librarykeywordrunner.py#L121

This functionality needs tests as well. To understand how to add them, you need to read the [README](https://github.com/robotframework/robotframework/blob/master/atest/README.rst). Then you need to add the runner test on the `atest/robot` side and the actual test on the `atest/testdata` side. Related suites are:

https://github.com/robotframework/robotframework/blob/master/atest/robot/cli/dryrun/executed_builtin_keywords.robot
https://github.com/robotframework/robotframework/blob/master/atest/testdata/cli/dryrun/executed_builtin_keywords.robot

On the runner side you can use keywords provided by [atest_resource.robot](https://github.com/robotframework/robotframework/blob/master/atest/resources/atest_resource.robot) for validating tags.

**Comment by pekkaklarck** [comment]
We are just starting RF 4.1 development and this issue could be included into it. The target for the release is early July and plan is to have a beta with all features out by the end of June.

**Comment by vinayvennela** [comment]
I'll start working on this

**Comment by pekkaklarck** [comment]
Great! I'll assign this to RF 4.1 scope.

**Comment by pekkaklarck** [comment]
This was implemented by PR #4007. Thanks for contribution @vinayvennela!
